python/src/trezorlib/authentication.py

```diff
@@ -0,0 +1,306 @@
+    return ec.EllipticCurvePublicKey.from_encoded_point(
+        ec.SECP256R1(), bytes.fromhex(pubkey_hex)
+    )
+
+
+CHALLENGE_HEADER = b"AuthenticateDevice:"
+
+
+class RootCertificate(t.NamedTuple):
+    name: str
+    device: str
+    devel: bool
+    pubkey: ec.EllipticCurvePublicKey
+
+
+ROOT_PUBLIC_KEYS = [
+    RootCertificate(
+        "Trezor Company",
+        "Trezor Safe 3",
+        False,
+        _pk_p256(
+            "04ca97480ac0d7b1e6efafe518cd433cec2bf8ab9822d76eafd34363b55d63e60"
+            "380bff20acc75cde03cffcb50ab6f8ce70c878e37ebc58ff7cca0a83b16b15fa5"
+        ),
+    ),
+    RootCertificate(
+        "Trezor Company",
+        "Trezor Safe 5",
+        False,
+        _pk_p256(
+            "041854b27fb1d9f65abb66828e78c9dc0ca301e66081ab0c6a4d104f9df1cd0ad"
+            "5a7c75f77a8c092f55cf825d2abaf734f934c9394d5e75f75a5a06a5ee9be93ae"
+        ),
+    ),
+    RootCertificate(
+        "TESTING ENVIRONMENT. DO NOT USE THIS DEVICE",
+        "Trezor Safe 3",
+        True,
+        _pk_p256(
+            "047f77368dea2d4d61e989f474a56723c3212dacf8a808d8795595ef38441427c"
+            "4389bc454f02089d7f08b873005e4c28d432468997871c0bf286fd3861e21e96a"
+        ),
+    ),
+    RootCertificate(
+        "TESTING ENVIRONMENT. DO NOT USE THIS DEVICE",
+        "Trezor Safe 5",
+        True,
+        _pk_p256(
+            "04e48b69cd7962068d3cca3bcc6b1747ef496c1e28b5529e34ad7295215ea161d"
+            "be8fb08ae0479568f9d2cb07630cb3e52f4af0692102da5873559e45e9fa72959"
+        ),
+    ),
+]
+
+
+class DeviceNotAuthentic(Exception):
+    pass
+
+
+class Certificate:
+    def __init__(self, cert_bytes: bytes) -> None:
+        self.cert_bytes = cert_bytes
+        self.cert = x509.load_der_x509_certificate(cert_bytes)
+
+    def __str__(self) -> str:
+        return self.cert.subject.rfc4514_string()
+
+    def public_key_bytes(self) -> bytes:
+        return self.cert.public_key().public_bytes(
+            serialization.Encoding.X962,
+            serialization.PublicFormat.UncompressedPoint,
+        )
+
+    def verify(self, signature: bytes, message: bytes) -> None:
+        cert_pubkey = self.cert.public_key()
+        assert isinstance(cert_pubkey, ec.EllipticCurvePublicKey)
+        cert_pubkey.verify(
+            self.fix_signature(signature),
+            message,
+            ec.ECDSA(hashes.SHA256()),
+        )
+
+    def verify_by(self, pubkey: ec.EllipticCurvePublicKey) -> None:
+        algo_params = self.cert.signature_algorithm_parameters
+        assert isinstance(algo_params, ec.ECDSA)
+        pubkey.verify(
+            self.fix_signature(self.cert.signature),
+            self.cert.tbs_certificate_bytes,
+            algo_params,
+        )
+
+    def verify_issued_by(self, issuer: "Certificate") -> None:
+        pubkey = issuer.cert.public_key()
+        assert isinstance(pubkey, ec.EllipticCurvePublicKey)
+        self.verify_by(pubkey)
```

Comment:
I think we should be checking more in this function.

Most importantly:
- The issuer's certificate should have the cA boolean value in basic constraints set to true.

Secondly:
- The pathLenConstraint should be checked.
- The certificate issuer's name should match the subject name in the issuer's certificate.

The full process is described in https://datatracker.ietf.org/doc/html/rfc5280#section-6.1, but we don't need to be as thorough as that.

Reply:
5623f6d6f7f5a8a1e49b73959117d1ec499bc579